diff -u -r a/contrib/pgcrypto/Makefile b/contrib/pgcrypto/Makefile
--- a/contrib/pgcrypto/Makefile	2014-08-06 11:33:07.000000000 +0200
+++ b/contrib/pgcrypto/Makefile	2014-08-06 14:15:18.781187060 +0200
@@ -26,7 +26,7 @@
 OBJS		= $(SRCS:.c=.o) $(WIN32RES)
 
 EXTENSION = pgcrypto
-DATA = pgcrypto--1.2.sql pgcrypto--1.0--1.1.sql pgcrypto--1.1--1.2.sql pgcrypto--unpackaged--1.0.sql
+DATA = pgcrypto--1.2.sql
 PGFILEDESC = "pgcrypto - cryptographic functions"
 
 REGRESS = init md5 sha1 hmac-md5 hmac-sha1 blowfish rijndael \
